src/object.rs

```diff
@@ -0,0 +1,321 @@
+use std::fmt;
+use std::collections::HashMap;
+use std::hash::{Hash,Hasher};
+use std::cell::RefCell;
+use std::rc::Rc;
+use ast;
+
+#[derive(Hash, Eq, PartialEq, Clone, Debug)]
+pub enum Object {
+    Int(i64),
+    Bool(bool),
+    String(String),
+    Return(Rc<Return>),
+    Function(Rc<Function>),
+    Builtin(Builtin),
+    Array(Rc<Array>),
+    Hash(Rc<MonkeyHash>),
+    Null,
+}
+
+impl Object {
+    pub fn inspect(&self) -> String {
+        match self {
+            Object::Int(i) => i.to_string(),
+            Object::Bool(b) => b.to_string(),
+            Object::String(s) => s.clone(),
+            Object::Return(r) => r.value.inspect(),
+            Object::Function(f) => f.inspect(),
+            Object::Builtin(b) => b.inspect(),
+            Object::Array(a) => a.inspect(),
+            Object::Hash(h) => h.inspect(),
+            Object::Null => String::from("null"),
+        }
+    }
+}
+impl fmt::Display for Object {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        write!(f, "{}", self.inspect())
+    }
+}
+
+#[derive(Eq, PartialEq, Clone, Debug)]
+pub struct MonkeyHash {
+    pub pairs: HashMap<Rc<Object>,Rc<Object>>,
+}
+impl MonkeyHash {
+    fn inspect(&self) -> String {
+        let pairs: Vec<String> = (&self.pairs).into_iter().map(|(key, value)| format!("{}: {}", key.inspect(), value.inspect())).collect();
+        format!("{{{}}}", pairs.join(", "))
+    }
+}
+impl Hash for MonkeyHash {
+    fn hash<H: Hasher>(&self, _state: &mut H) {
+        // should never happen
+        panic!("hash not implmented for monkey hash");
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Array {
+    pub elements: Vec<Rc<Object>>,
+}
+
+impl Array {
+    fn inspect(&self) -> String {
+        let elements: Vec<String> = (&self.elements).into_iter().map(|e| e.to_string()).collect();
+        format!("[{}]", elements.join(", "))
+    }
+}
+impl PartialEq for Array {
+    fn eq(&self, _other: &Array) -> bool {
+        // TODO: implment this later, but this shouldn't get used right now
+        panic!("eq not implemented for array");
+    }
+}
+impl Eq for Array {}
+impl Hash for Array {
+    fn hash<H: Hasher>(&self, _state: &mut H) {
+        // we should never hash an array so should be fine
+        panic!("hash for array not supported");
+    }
+}
+
+#[derive(Hash, Eq, PartialEq, Clone, Debug)]
+pub enum Builtin {
+    Len,
+    First,
+    Last,
+    Rest,
+    Push,
+    Puts,
+}
+
+impl Builtin {
+    pub fn lookup(name: &str) -> Option<Object> {
+        match name {
+            "len" => Some(Object::Builtin(Builtin::Len)),
```

Comment:
Same here as with `Object::Null`, would be nice to re-use a constant of some sort. How to do this?

Reply:
You only need `lazy_static` if the value can't be computed at compile-time.
Btw, the range of expressions that can be computed at compile-time is being expanded (through `const fn`, similar to const-expr in C++).

For this case, you could just define a non-lazy (normal) static or const:

`const BUILTIN_LEN = Object::Builtin(Builtin::Len);`

The difference between a const and a static is that consts don't have a memory location, they are inlined at every use location, so it's not recommended to use const for arrays, they will be copied onto the stack at the use locations and can blow the stack.